File: app/runtime/actor.py
"""Actor：统一单循环执行 plan 和 atomic 两种 task。

工具集来自 ctx.resources（kind="tool"），由 Reasoner 按 task.type 差异化加载：
  - plan task：submit_plan
  - atomic task：外部工具 + request_human_input

循环由 ControlSignal.TASK_COMPLETE（plan 路径）和"无工具调用"（act 路径）统一终止。
身份与协议文本由 soul 模板承载，Actor 不感知 plan/act 模式。
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from app.config.settings import get_settings
from app.domain.models.agent import Agent
from app.domain.models.session import Session
from app.domain.models.task import Task
from app.domain.services.session_service import SessionService
from app.domain.services.task_service import TaskService
from app.llm.base import BaseChatClient, LLMMessage
from app.runtime.prompt_builder import ActorPromptBuilder, PromptBuilderFactory
from app.runtime.types import ActorResult, ContextResource, ConversationTurn, ReasoningContext, ToolCallRecord
from app.tools.types import CallContext, ToolResult

if TYPE_CHECKING:
    from app.runtime.tool_gateway import ToolGateway

logger = logging.getLogger(__name__)



class Actor:
    """统一 LLM 执行入口：plan 和 act 共用同一个 tool-use 循环。"""

    def __init__(
        self,
        tool_gateway: "ToolGateway",
        task_svc: TaskService,
        session_svc: SessionService | None = None,
        prompt_builder: ActorPromptBuilder | None = None,
    ) -> None:
        self._tool_gateway = tool_gateway
        self._task_svc = task_svc
        self._session_svc = session_svc
        self._prompt_builder = prompt_builder or PromptBuilderFactory.for_actor()

    def act(self, task: Task, ctx: ReasoningContext, agent: Agent, session: Session) -> ActorResult:
        """执行单个 task"""
        if task.status == "PENDING":
            self._task_svc.transition(task.id, "ACTIVE", task.session_id)

        system_prompt = self._prompt_builder.build_system_prompt(ctx)
        messages = self._prompt_builder.build_messages(task, ctx)
        tools = [r.llm_tool for r in ctx.actor_resources if r.kind == "tool" and r.llm_tool is not None]

        session_id = getattr(task, "session_id", "")
        _wd = (
            (task.settings.get("working_dir") if task.settings else None)
            or getattr(agent, "settings", {}).get("working_dir")
            or get_settings().bash_exec_cwd
        )
        from app.config.settings import resolve_working_dir
        toolcall_ctx = CallContext(session_id=session_id, agent_id=agent.id, task=task, working_dir=resolve_working_dir(_wd or ""))
        llm_client = self._resolve_llm_client(session)
        _sse = self._get_sse(session_id)
        is_daemon = bool(task.settings.get("_daemon") if task.settings else False)

        tool_calls_made: list[ToolCallRecord] = []
        conversation_turns: list[ConversationTurn] = []
        last_text = ""
        max_context_tokens = 0
        last_prompt_tokens = ctx.token_estimate
        exit_reason = "max_rounds"

        for _round in range(agent.loop_guard.actor_max_tool_rounds):
            messages = self._prompt_builder.sanitize_messages(messages)
            messages_sent = list(messages)
            self._push_prompt_event(_sse, session_id, _round, system_prompt, messages, ctx, is_daemon)

            round_max_tokens = _compute_max_tokens(llm_client, last_prompt_tokens)
            full_text, reasoning_text, tool_call_acc, image_acc, _usage = self._stream_llm(
                llm_client, messages, system_prompt, tools, _round, _sse, session_id, round_max_tokens, is_daemon
            )
            if _usage:
                if self._session_svc and session_id:
                    self._session_svc.add_tokens(
                        session_id,
                        input_tokens=_usage.prompt_tokens or 0,
                        output_tokens=_usage.completion_tokens or 0,
                        context_tokens=_usage.prompt_tokens or 0,
                    )
                if _usage.prompt_tokens:
                    max_context_tokens = max(max_context_tokens, _usage.prompt_tokens)
            last_text = full_text
            parsed = llm_client.parse_stream_acc(full_text, tool_call_acc, images=image_acc)
            tool_calls_from_stream = parsed.tool_calls

            round_tool_calls: list[ToolCallRecord] = []
            done = False

            if tool_calls_from_stream:
                messages = self._prompt_builder.append_assistant_tool_calls(
                    messages, full_text, tool_calls_from_stream, reasoning_text or None
                )
                round_tool_calls, messages, done = self._execute_tools(
                    tool_calls_from_stream, agent, task, toolcall_ctx, messages
                )
                tool_calls_made.extend(round_tool_calls)

            if _usage and _usage.prompt_tokens:
                from app.common.utils import estimate_tokens
                tool_text = " ".join(tc.result for tc in round_tool_calls)
                last_prompt_tokens = _usage.prompt_tokens + estimate_tokens(tool_text)

            conversation_turns.append(ConversationTurn(
                round=_round,
                messages_sent=messages_sent,
                llm_text=last_text,
                tool_calls=round_tool_calls,
                images=parsed.images,
            ))

            context_limit_hit = (
                _usage is not None
                and _usage.prompt_tokens is not None
                and llm_client.context_limit > 0
                and _usage.prompt_tokens >= llm_client.context_limit * 0.8
            )
            if context_limit_hit:
                logger.warning(
                    "Actor context limit reached: prompt_tokens=%s >= context_limit=%s, stopping loop",
                    _usage.prompt_tokens, llm_client.context_limit,
                )

            if not tool_calls_from_stream or done or context_limit_hit:
                exit_reason = "context_limit" if context_limit_hit else "normal"
                break

        return self._build_result(task, tool_calls_made, conversation_turns, last_text, max_context_tokens, exit_reason)

    # ── 私有辅助方法 ──────────────────────────────────────────────────────────

    def _resolve_llm_client(self, session: Session) -> BaseChatClient:
        from app.config.settings import get_settings
        from app.llm.registry import get_llm_registry
        provider = session.llm_provider or get_settings().default_llm_provider
        return get_llm_registry().get_client(provider, session.llm_model or None)

    def _get_sse(self, session_id: str):
        if not session_id:
            return None
        try:
            from app.common.sse_bus import get_sse_bus
            return get_sse_bus()
        except Exception:
            return None

    def _sse_push(self, _sse, session_id: str, event: dict) -> None:
        if _sse:
            try:
                _sse.push(session_id, event)
            except Exception:
                pass

    def _push_prompt_event(self, _sse, session_id: str, _round: int, system_prompt: str, messages, ctx: ReasoningContext, is_daemon: bool = False) -> None:
        self._sse_push(_sse, session_id, {
            "type": "daemon_prompt" if is_daemon else "llm_prompt",
            "source": "actor",
            "round_label": f"actor_round_{_round}",
            "task_id": ctx.current_task.id if ctx.current_task else "",
            "agent_id": ctx.current_task.assigned_agent_id if ctx.current_task else "",
            "system_prompt": system_prompt,
            "messages": [{"role": m.role, "content": m.content} for m in messages],
            "tool_names": [r.name for r in ctx.actor_resources if r.kind == "tool" and r.llm_tool is not None],
        })

    def _stream_llm(self, llm_client: BaseChatClient, messages, system_prompt, tools, _round: int, _sse, session_id: str, max_tokens: int | None = None, is_daemon: bool = False):
        full_text = ""
        reasoning_text = ""
        tool_call_acc: dict[int, dict] = {}
        image_acc: list = []
        finish_reason = None
        final_usage = None

        from app.common.errors import AppError
        try:
            for chunk in llm_client.stream_message(messages=messages, system_prompt=system_prompt, tools=tools, max_tokens=max_tokens):
                if chunk.text_delta:
                    full_text += chunk.text_delta
                    if not is_daemon:
                        self._sse_push(_sse, session_id, {"type": "text_delta", "delta": chunk.text_delta, "round": _round})
                if chunk.reasoning_delta:
                    reasoning_text += chunk.reasoning_delta
                    if not is_daemon:
                        self._sse_push(_sse, session_id, {
                            "type": "reasoning_delta",
                            "delta": chunk.reasoning_delta,
                            "round": _round,
                        })
                if chunk.tool_call_delta:
                    tool_call_acc[chunk.tool_call_delta["index"]] = chunk.tool_call_delta
                if chunk.image:
                    image_acc.append(chunk.image)
                    if not is_daemon:
                        self._sse_push(_sse, session_id, {
                            "type": "image",
                            "media_type": chunk.image.media_type,
                            "source_type": chunk.image.source_type,
                            "data": chunk.image.data,
                            "round": _round,
                        })
                if chunk.is_done:
                    finish_reason = chunk.finish_reason
                    final_usage = chunk.usage
                    if chunk.error:
                        raise AppError("LLM_API_ERROR", str(chunk.error))
        except AppError:
            raise
        except Exception as e:
            raise AppError("LLM_API_ERROR", str(e)) from e

        if final_usage is None and not full_text and not tool_call_acc:
            raise AppError("LLM_API_ERROR", "LLM stream ended without completion")

        if not is_daemon and reasoning_text:
            self._sse_push(_sse, session_id, {
                "type": "reasoning_done",
                "text": reasoning_text,
                "round": _round,
            })
        self._sse_push(_sse, session_id, {
            "type": "daemon_message" if is_daemon else "text_done",
            "text": full_text,
            "round": _round,
            "context_tokens": final_usage.prompt_tokens if final_usage else None,
        })
        logger.info(
            "Actor LLM stream finished: round=%s finish_reason=%s usage=%s text_len=%s reasoning_len=%s tool_calls=%s images=%s",
            _round,
            finish_reason,
            final_usage,
            len(full_text),
            len(reasoning_text),
            len(tool_call_acc),
            len(image_acc),
        )
        return full_text, reasoning_text, tool_call_acc, image_acc, final_usage

    def _execute_tools(self, tool_calls, agent: Agent, task: Task, toolcall_ctx: CallContext, messages):
        round_tool_calls: list[ToolCallRecord] = []
        done = False

        for tool_call in tool_calls:
            try:
                result = self._tool_gateway.call(
                    tool_name=tool_call.name,
                    arguments=tool_call.input,
                    capability=agent.actor,
                    task_id=task.id,
                    ctx=toolcall_ctx,
                )
            except Exception as e:
                logger.warning("Actor: tool '%s' raised %s", tool_call.name, e)
                result = ToolResult(content=str(e), is_error=True)

            if task.actor_done:
                done = True

            from app.llm.types import content_to_text
            result_text = content_to_text(result.content) if isinstance(result.content, list) else (result.content or "")
            record = ToolCallRecord(
                tool_name=tool_call.name,
                arguments=tool_call.input,
                result=result_text,
                is_error=result.is_error,
                tool_call_id=tool_call.id,
            )
            round_tool_calls.append(record)
            messages = self._prompt_builder.append_tool_result(
                messages, tool_call.name, result, tool_call_id=tool_call.id
            )

        return round_tool_calls, messages, done

    # ── ActorResult 构建 ───────────────────────────────────────────────────

    def _build_result(
        self,
        task: Task,
        tool_calls_made: list[ToolCallRecord],
        conversation_turns: list[ConversationTurn],
        last_text: str,
        context_tokens: int = 0,
        exit_reason: str = "normal",
    ) -> ActorResult:
        """plan 和 act 统一走文本路径；task 创建由 Observer 阶段负责。"""
        skill_used = task.settings.get("skill_name")
        actor_mode = "skill" if skill_used else ("tool_use" if tool_calls_made else "text")

        error_parts: list[str] = []
        failed_tools = [tc for tc in tool_calls_made if tc.is_error]
        if failed_tools:
            summaries = [f"{tc.tool_name}: {tc.result}" for tc in failed_tools]
            error_parts.append("Tool errors: " + "; ".join(summaries))
        if exit_reason == "max_rounds":
            error_parts.append("Actor reached max tool rounds without completion")
        elif exit_reason == "context_limit":
            error_parts.append("Actor stopped: context limit reached")

        return ActorResult(
            task_id=task.id,
            success=exit_reason not in ("max_rounds", "context_limit"),
            output=last_text,
            tool_calls_made=tool_calls_made,
            conversation_turns=conversation_turns,
            actor_mode=actor_mode,
            skill_used=skill_used,
            error="\n".join(error_parts) if error_parts else None,
            context_tokens=context_tokens,
            exit_reason=exit_reason,
        )


def _compute_max_tokens(llm_client: BaseChatClient, prompt_tokens: int) -> int:
    available = llm_client.context_limit - prompt_tokens
    return max(1, min(llm_client.max_output_tokens, available))

File: miniagents-assembled/src/main/java/com/codex/miniagents/runtime/Actor.java
package com.codex.miniagents.runtime;

import com.codex.miniagents.common.SseBus;
import com.codex.miniagents.config.MiniAgentsProperties;
import com.codex.miniagents.domain.model.agent.Agent;
import com.codex.miniagents.domain.model.task.Task;
import com.codex.miniagents.domain.model.task.TaskStatus;
import com.codex.miniagents.domain.service.TaskService;
import com.codex.miniagents.llm.ChatClient;
import com.codex.miniagents.llm.model.LlmMessage;
import com.codex.miniagents.llm.model.LlmTool;
import com.codex.miniagents.llm.model.StreamChunk;
import com.codex.miniagents.llm.model.ToolCallBlock;
import com.codex.miniagents.llm.registry.LlmClientProvider;
import com.codex.miniagents.runtime.model.ActorResult;
import com.codex.miniagents.runtime.model.ContextResource;
import com.codex.miniagents.runtime.model.ConversationTurn;
import com.codex.miniagents.runtime.model.ReasoningContext;
import com.codex.miniagents.runtime.model.ToolCallRecord;
import com.codex.miniagents.runtime.prompt.ActorPromptBuilder;
import com.codex.miniagents.runtime.prompt.PromptBuilderFactory;
import com.codex.miniagents.tools.model.CallContext;
import com.codex.miniagents.tools.model.ToolResult;

import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.time.Instant;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Component
public class Actor {
    private final LlmClientProvider llmClientProvider;

    private final ToolGateway toolGateway;

    private final TaskService taskService;

    private final MiniAgentsProperties properties;

    public Actor(LlmClientProvider llmClientProvider, ToolGateway toolGateway,
        TaskService taskService, MiniAgentsProperties properties) {
        this.llmClientProvider = llmClientProvider;
        this.toolGateway = toolGateway;
        this.taskService = taskService;
        this.properties = properties;
    }

    public ActorResult act(Task task, ReasoningContext context, Agent agent) {
        if (task.getStatus() == TaskStatus.PENDING) {
            taskService.transition(task.getId(), TaskStatus.ACTIVE);
        }

        ActorPromptBuilder promptBuilder = PromptBuilderFactory.forActor();
        ChatClient llmClient = resolveLlmClient(agent);
        String systemPrompt = promptBuilder.buildSystemPrompt(context);
        List<LlmMessage> messages = promptBuilder.buildMessages(task, context);
        List<LlmTool> tools = resolveTools(context);

        List<ToolCallRecord> toolCalls = new ArrayList<>();
        List<ConversationTurn> conversationTurns = new ArrayList<>();
        String lastText = "";
        String sessionId = task.getSessionId();

        int maxRounds = 1;
        if (agent.getLoopGuard() != null) {
            maxRounds = Math.max(1, agent.getLoopGuard().getActorMaxToolRounds());
        }

        for (int round = 0; round < maxRounds; round++) {
            List<LlmMessage> messagesSent = cloneMessages(messages);
            pushPromptEvent(sessionId, round, systemPrompt, messages, context);

            StreamRoundResult streamResult = streamLlm(llmClient, messages, systemPrompt, tools, round, sessionId);
            String fullText = streamResult.fullText;
            Map<Integer, Map<String, Object>> toolCallAcc = streamResult.toolCallAcc;

            List<ToolCallBlock> parsedToolCalls = promptBuilder.buildToolCallsFromStream(toolCallAcc);
            if (parsedToolCalls.isEmpty()) {
                lastText = fullText;
                conversationTurns.add(ConversationTurn.builder()
                    .round(round)
                    .messagesSent(messagesSent)
                    .llmText(defaultString(lastText))
                    .toolCalls(List.of())
                    .build());
                break;
            }

            messages = promptBuilder.appendAssistantToolCalls(messages, defaultString(fullText), parsedToolCalls);

            ToolExecutionResult executionResult = executeTools(promptBuilder, parsedToolCalls, agent, task, messages,
                sessionId);
            messages = executionResult.messages;
            List<ToolCallRecord> roundToolCalls = executionResult.roundToolCalls;
            boolean done = executionResult.done;
            toolCalls.addAll(roundToolCalls);
            lastText = fullText;
            conversationTurns.add(ConversationTurn.builder()
                .round(round)
                .messagesSent(messagesSent)
                .llmText(defaultString(lastText))
                .toolCalls(roundToolCalls)
                .build());
            if (done) {
                break;
            }
        }

        Map<String, Object> settings = task.getSettings() == null ? Map.of() : task.getSettings();
        String skillUsed = settings.get("skill_name") == null ? null : String.valueOf(settings.get("skill_name"));
        boolean hasSkill = skillUsed != null && !skillUsed.isBlank();
        String actorMode = hasSkill
            ? "skill"
            : (toolCalls.isEmpty() ? "text" : "tool_use");
        return ActorResult.builder()
            .taskId(task.getId())
            .success(true)
            .output(lastText == null ? "" : lastText)
            .toolCallsMade(toolCalls)
            .conversationTurns(conversationTurns)
            .actorMode(actorMode)
            .skillUsed(hasSkill ? skillUsed : null)
            .build();
    }

    private ChatClient resolveLlmClient(Agent agent) {
        return llmClientProvider.get(agent.getLlmName(), agent.getLlmModel());
    }

    private List<LlmTool> resolveTools(ReasoningContext context) {
        return context.getActorResources() == null
            ? List.of()
            : context.getActorResources().stream()
                .filter(r -> "tool".equals(r.getKind()) && r.getLlmTool() != null)
                .map(ContextResource::getLlmTool)
                .collect(Collectors.toList());
    }

    private void pushPromptEvent(String sessionId, int round, String systemPrompt, List<LlmMessage> messages,
        ReasoningContext context) {
        try {
            SseBus.getInstance().push(sessionId, Map.of(
                "type", "llm_prompt",
                "source", "actor",
                "round_label", "actor_round_" + round,
                "system_prompt", systemPrompt,
                "messages", messages.stream().map(m -> {
                    Map<String, Object> msg = new LinkedHashMap<>();
                    msg.put("role", defaultString(m.getRole()));
                    msg.put("content", defaultString(m.getContent()));
                    if (m.getToolCallId() != null && !m.getToolCallId().isBlank()) {
                        msg.put("tool_call_id", m.getToolCallId());
                    }
                    if (m.getToolCalls() != null && !m.getToolCalls().isEmpty()) {
                        msg.put("tool_calls", m.getToolCalls());
                    }
                    return msg;
                }).toList(),
                "tool_names", context.getActorResources() == null ? List.of() : context.getActorResources().stream()
                    .filter(r -> "tool".equals(r.getKind()) && r.getLlmTool() != null)
                    .map(r -> r.getLlmTool().getName())
                    .toList()
            ));
        } catch (Exception ignored) {
        }
    }

    private StreamRoundResult streamLlm(ChatClient llmClient, List<LlmMessage> messages, String systemPrompt,
        List<LlmTool> tools, int round, String sessionId) {
        String fullText = "";
        Map<Integer, Map<String, Object>> toolCallAcc = new LinkedHashMap<>();
        for (Iterator<StreamChunk> stream = llmClient.streamMessage(messages, systemPrompt, tools); stream.hasNext();) {
            StreamChunk chunk = stream.next();
            if (chunk.getTextDelta() != null && !chunk.getTextDelta().isEmpty()) {
                fullText += chunk.getTextDelta();
                pushSseEvent(sessionId, Map.of(
                    "type", "text_delta",
                    "delta", chunk.getTextDelta(),
                    "round", round
                ));
            }
            if (chunk.getToolCallDelta() != null && !chunk.getToolCallDelta().isEmpty()) {
                Object idxObj = chunk.getToolCallDelta().get("index");
                int idx = idxObj instanceof Number n ? n.intValue() : 0;
                toolCallAcc.put(idx, new LinkedHashMap<>(chunk.getToolCallDelta()));
            }
        }
        pushSseEvent(sessionId, Map.of(
            "type", "text_done",
            "text", fullText,
            "round", round
        ));
        return new StreamRoundResult(fullText, toolCallAcc);
    }

    private ToolExecutionResult executeTools(ActorPromptBuilder promptBuilder, List<ToolCallBlock> parsedToolCalls,
        Agent agent, Task task, List<LlmMessage> messages, String sessionId) {
        CallContext callContext = CallContext.builder()
            .sessionId(sessionId == null ? "" : sessionId)
            .agentId(agent == null || agent.getId() == null ? "" : agent.getId())
            .agent(agent)
            .task(task)
            .workingDir(resolveWorkingDir(task, agent))
            .build();
        List<ToolCallRecord> roundToolCalls = new ArrayList<>();
        boolean done = false;
        List<LlmMessage> nextMessages = messages;
        for (ToolCallBlock toolCall : parsedToolCalls) {
            Map<String, Object> args = toolCall.getInput() == null ? Map.of() : toolCall.getInput();
            ToolResult result;
            try {
                result = toolGateway.call(toolCall.getName(), args, agent, task.getId(), callContext);
            } catch (Exception e) {
                result = ToolResult.builder().content(String.valueOf(e.getMessage())).isError(true)
                    .errorCode("TOOL_EXEC_ERROR").build();
            }
            if (isTaskCompleteSignal(result) || task.isActorDone()) {
                done = true;
            }
            ToolCallRecord record = ToolCallRecord.builder()
                .toolCallId(toolCall.getId())
                .toolName(toolCall.getName())
                .arguments(args)
                .result(result.getContent() == null ? "" : result.getContent())
                .isError(result.isError())
                .build();
            roundToolCalls.add(record);
            nextMessages = promptBuilder.appendToolResult(nextMessages, toolCall.getName(), result, toolCall.getId());
            pushSseEvent(sessionId, Map.of(
                "type", "tool_call",
                "tool_name", toolCall.getName(),
                "arguments", args,
                "result", result.getContent() == null ? "" : result.getContent(),
                "is_error", result.isError(),
                "created_at", Instant.now().toString()
            ));
        }
        return new ToolExecutionResult(roundToolCalls, nextMessages, done);
    }

    private boolean isTaskCompleteSignal(ToolResult result) {
        if (result == null || result.getMetadata() == null) {
            return false;
        }
        Object signal = result.getMetadata().get("control_signal");
        return ControlSignal.TASK_COMPLETE.name().equals(String.valueOf(signal));
    }

    private List<LlmMessage> cloneMessages(List<LlmMessage> source) {
        List<LlmMessage> copy = new ArrayList<>();
        for (LlmMessage m : source) {
            copy.add(LlmMessage.builder()
                .role(m.getRole())
                .content(m.getContent())
                .toolCallId(m.getToolCallId())
                .toolCalls(m.getToolCalls())
                .build());
        }
        return copy;
    }

    private String defaultString(String value) {
        return value == null ? "" : value;
    }

    private void pushSseEvent(String sessionId, Map<String, Object> event) {
        try {
            SseBus.getInstance().push(sessionId, event);
        } catch (Exception ignored) {
        }
    }

    private String resolveWorkingDir(Task task, Agent agent) {
        if (task != null && task.getSettings() != null) {
            Object value = task.getSettings().get("working_dir");
            if (value != null) {
                String workingDir = String.valueOf(value);
                if (!workingDir.isBlank()) {
                    return workingDir;
                }
            }
        }
        if (agent != null && agent.getSettings() != null) {
            Object value = agent.getSettings().get("working_dir");
            if (value != null) {
                String workingDir = String.valueOf(value);
                if (!workingDir.isBlank()) {
                    return workingDir;
                }
            }
        }
        return properties.getBashExecCwd() == null ? "" : properties.getBashExecCwd();
    }

    private record StreamRoundResult(
        String fullText,
        Map<Integer, Map<String, Object>> toolCallAcc
    ) {
    }

    private record ToolExecutionResult(
        List<ToolCallRecord> roundToolCalls,
        List<LlmMessage> messages,
        boolean done
    ) {
    }
}
